sw/device/lib/crypto/drivers/entropy.c

```diff
@@ -107,6 +269,194 @@ static status_t csrng_send_app_cmd(uint32_t reg_address,
+  // command can get in after the clear above.
+  abs_mmio_write32(edn_address + EDN_CTRL_REG_OFFSET, EDN_CTRL_REG_RESVAL);
+}
+
+/**
+ * Blocks until EDN instance is ready to execute a new CSNRG command.
+ *
+ * @param edn_address EDN base address.
+ * @returns an error if the EDN error status bit is set.
+ */
+OT_WARN_UNUSED_RESULT
+static status_t edn_ready_block(uint32_t edn_address) {
+  uint32_t reg;
+  do {
+    reg = abs_mmio_read32(edn_address + EDN_SW_CMD_STS_REG_OFFSET);
+  } while (!bitfield_bit32_read(reg, EDN_SW_CMD_STS_CMD_RDY_BIT));
+
+  if (bitfield_bit32_read(reg, EDN_SW_CMD_STS_CMD_STS_BIT)) {
+    return INTERNAL();
+  }
+  return OK_STATUS();
+}
+
+/**
+ * Configures EDN instance based on `config` options.
+ *
+ * @param config EDN configuration options.
+ * @returns error on failure.
+ */
+OT_WARN_UNUSED_RESULT
+static status_t edn_configure(const edn_config_t *config) {
+  TRY(csrng_send_app_cmd(config->base_address + EDN_RESEED_CMD_REG_OFFSET,
+                         config->reseed));
+  TRY(csrng_send_app_cmd(config->base_address + EDN_GENERATE_CMD_REG_OFFSET,
+                         config->generate));
+  abs_mmio_write32(
+      config->base_address + EDN_MAX_NUM_REQS_BETWEEN_RESEEDS_REG_OFFSET,
+      config->reseed_interval);
+
+  uint32_t reg =
+      bitfield_field32_write(0, EDN_CTRL_EDN_ENABLE_FIELD, kMultiBitBool4True);
+  reg = bitfield_field32_write(reg, EDN_CTRL_AUTO_REQ_MODE_FIELD,
+                               kMultiBitBool4True);
+  abs_mmio_write32(config->base_address + EDN_CTRL_REG_OFFSET, reg);
+
+  TRY(edn_ready_block(config->base_address));
+  TRY(csrng_send_app_cmd(config->base_address + EDN_SW_CMD_REQ_REG_OFFSET,
+                         config->instantiate));
+  return edn_ready_block(config->base_address);
+}
+
+/**
+ * Stops the current mode of operation and disables the entropy_src module.
+ *
+ * All configuration registers are set to their reset values to avoid
+ * synchronization issues with internal FIFOs.
+ */
+static void entropy_src_stop(void) {
+  abs_mmio_write32(kBaseEntropySrc + ENTROPY_SRC_MODULE_ENABLE_REG_OFFSET,
+                   ENTROPY_SRC_MODULE_ENABLE_REG_RESVAL);
+
+  // Set default values for other critical registers to avoid synchronization
+  // issues.
+  abs_mmio_write32(kBaseEntropySrc + ENTROPY_SRC_ENTROPY_CONTROL_REG_OFFSET,
+                   ENTROPY_SRC_ENTROPY_CONTROL_REG_RESVAL);
+  abs_mmio_write32(kBaseEntropySrc + ENTROPY_SRC_CONF_REG_OFFSET,
+                   ENTROPY_SRC_CONF_REG_RESVAL);
+  abs_mmio_write32(kBaseEntropySrc + ENTROPY_SRC_HEALTH_TEST_WINDOWS_REG_OFFSET,
+                   ENTROPY_SRC_HEALTH_TEST_WINDOWS_REG_RESVAL);
+  abs_mmio_write32(kBaseEntropySrc + ENTROPY_SRC_ALERT_THRESHOLD_REG_OFFSET,
+                   ENTROPY_SRC_ALERT_THRESHOLD_REG_RESVAL);
+}
+
+/**
+ * Disables the entropy complex.
+ *
+ * The order of operations is important to avoid synchronization issues across
+ * blocks.
```

Comment:
nit: would you mind adding a little more detail here? What kind of synchronization issues are you protecting against, and is there any information about them in the CSRNG/EDN docs?

Reply:
Done.